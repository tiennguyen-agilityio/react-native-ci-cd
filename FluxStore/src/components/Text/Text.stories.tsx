import {Meta, StoryObj} from '@storybook/react';

import Text from '.';

const meta: Meta<typeof Text> = {
  title: 'Components/Text',
  component: Text,
};

export default meta;

type Story = StoryObj<typeof Text>;

const props = {
  children: 'Lorem ipsum dolor sit amet',
};

export const Default: Story = {
  args: props,
};

export const Heading: Story = {
  args: {
    ...props,
    variant: 'heading',
  },
};

export const Title: Story = {
  args: {
    ...props,
    variant: 'title',
  },
};

export const SubTitle: Story = {
  args: {
    ...props,
    variant: 'subTitle',
  },
};

export const Description: Story = {
  args: {
    ...props,
    variant: 'description',
  },
};
