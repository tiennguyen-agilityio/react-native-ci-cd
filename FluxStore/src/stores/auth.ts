import AsyncStorage from '@react-native-async-storage/async-storage';
import {persist, createJSONStorage} from 'zustand/middleware';
import {createWithEqualityFn} from 'zustand/traditional';
import {shallow} from 'zustand/shallow';

// Interfaces
import {User} from '@/interfaces';

// Constants
import {STORAGE_KEYS} from '@/constants';

type States = {
  isAuthenticated?: boolean;
  authHydrated?: boolean;
  user: User | null;
};

type Actions = {
  setIsAuthenticated: (isAuthenticated: boolean) => void;
  setAuthHydrated: (hydrated: boolean) => void;
  setUser: (user: User) => void;
};

const INITIAL_STATE: States = {
  isAuthenticated: false,
  authHydrated: false,
  user: null,
};

export const useAuthStore = createWithEqualityFn<States & Actions>()(
  persist(
    set => ({
      ...INITIAL_STATE,
      setAuthHydrated: hydrated => set({authHydrated: hydrated}),
      setIsAuthenticated: (isAuthenticated: boolean) => {
        set({isAuthenticated});
      },
      setUser: (user: User) => {
        set({user});
      },
    }),
    {
      name: STORAGE_KEYS.AUTH,
      storage: createJSONStorage(() => AsyncStorage),
      onRehydrateStorage: () => state => {
        state?.setAuthHydrated(true);
      },
      partialize: state => {
        return {
          user: state.user,
          isAuthenticated: state.isAuthenticated,
        };
      },
    },
  ),
  shallow,
);
