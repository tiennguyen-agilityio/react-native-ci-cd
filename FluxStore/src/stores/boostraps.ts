import {createWithEqualityFn} from 'zustand/traditional';
import {persist, createJSONStorage} from 'zustand/middleware';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {shallow} from 'zustand/shallow';

// Constants
import {STORAGE_KEYS} from '@/constants';

type States = {
  isFirstLoad: boolean;
  bootHydrated: boolean;
};

type Actions = {
  finishBoot: () => void;
  setBootHydrated: (hydrated: boolean) => void;
};

const INITIAL_STATE: States = {
  isFirstLoad: true,
  bootHydrated: false,
};

export const useBootstrapsStore = createWithEqualityFn<States & Actions>()(
  persist(
    set => ({
      ...INITIAL_STATE,
      finishBoot: () => set({isFirstLoad: false}),
      setBootHydrated: hydrated => set({bootHydrated: hydrated}),
    }),
    {
      name: STORAGE_KEYS.FIRST_LOAD,
      storage: createJSONStorage(() => AsyncStorage),
      partialize: state => ({
        ...state,
        isFirstLoad: state.isFirstLoad,
      }),
      onRehydrateStorage: () => state => {
        state?.setBootHydrated(true);
      },
    },
  ),
  shallow,
);
