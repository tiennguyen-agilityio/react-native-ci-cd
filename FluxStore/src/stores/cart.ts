import {createWithEqualityFn} from 'zustand/traditional';
import {createJSONStorage, persist} from 'zustand/middleware';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {shallow} from 'zustand/shallow';

// Interfaces
import {Cart} from '@/interfaces';
import {STORAGE_KEYS} from '@/constants';

// Utils
import {calcTotalPrice} from '@/utils';

type States = {
  carts: Cart[] | [];
  totalPrice: string | number;
};

type Actions = {
  setCarts: (carts: Cart[]) => void;
  addNewCart: (cart: Cart) => void;
  updateCartItem: (cart: Cart) => void;
  removeCartItem: (id: string) => void;
  clearCart: () => void;
};

const INITIAL_STATE: States = {
  carts: [],
  totalPrice: 0,
};

export const useCartStore = createWithEqualityFn<States & Actions>()(
  persist(
    set => ({
      ...INITIAL_STATE,
      setCarts: (carts: Cart[]) => set({carts}),
      addNewCart: (cart: Cart) => {
        set(state => {
          const indexItem = state?.carts?.findIndex(item => item.id === cart.id);

          const carts =
            indexItem === -1
              ? [cart, ...state.carts]
              : state.carts.map((item, index) =>
                  index === indexItem ? {...item, quantity: item.quantity + cart.quantity} : item,
                );

          return {carts: carts, totalPrice: calcTotalPrice(carts)};
        });
      },
      updateCartItem: ({id, isChecked, quantity}) => {
        set(state => {
          const carts = state.carts.map(item =>
            item?.id === id ? {...item, quantity, isChecked} : item,
          );

          return {carts, totalPrice: calcTotalPrice(carts)};
        });
      },
      removeCartItem: (id: string) =>
        set(state => {
          const carts = state.carts.filter(c => c.id !== id);
          return {carts, totalPrice: calcTotalPrice(carts)};
        }),
      clearCart: () => set({...INITIAL_STATE}),
    }),
    {
      name: STORAGE_KEYS.CART,
      storage: createJSONStorage(() => AsyncStorage),

      // Ensure totalPrice is correct when data is restored
      onRehydrateStorage: () => state => {
        if (state) {
          state.totalPrice = calcTotalPrice(state.carts);
        }
      },
    },
  ),
  shallow,
);
