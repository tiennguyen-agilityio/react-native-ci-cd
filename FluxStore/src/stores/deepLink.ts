import {createWithEqualityFn} from 'zustand/traditional';
import {shallow} from 'zustand/shallow';

import {SCREENS} from '@/interfaces';
import {parseDeepLink} from '@/configs';

type PendingDeepLink = {
  stack?: SCREENS;
  screen?: SCREENS;
  params?: {
    id?: string;
  };
} | null;

type DeepLinkStore = {
  pendingDeepLink: PendingDeepLink;
  setPendingDeepLink: (url: string | null) => void;
};

export const useDeepLinkStore = createWithEqualityFn<DeepLinkStore>()(
  set => ({
    pendingDeepLink: null,
    setPendingDeepLink: (url: string | null) => {
      if (!url) {
        return set({pendingDeepLink: null});
      }
      const value = parseDeepLink(url);
      set({pendingDeepLink: value});
    },
  }),
  shallow,
);
