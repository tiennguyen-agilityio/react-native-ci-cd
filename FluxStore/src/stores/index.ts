export * from './boostraps';
export * from './theme';
export * from './auth';
export * from './cart';
export * from './deepLink';
