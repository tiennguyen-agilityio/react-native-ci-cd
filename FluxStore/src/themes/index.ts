export * from './colors';
export * from './typography';
export * from './metrics';
export * from './images';
export * from './theme';
