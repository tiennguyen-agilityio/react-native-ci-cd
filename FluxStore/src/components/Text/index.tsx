import {ReactNode, memo, useMemo} from 'react';
import {
  ColorValue,
  Text as RNText,
  TextProps as RNTextProps,
  StyleProp,
  StyleSheet,
  TextStyle,
} from 'react-native';

// Themes
import {fontSizes, fontWeights, letterSpacings} from '@/themes';
import {useThemeStore} from '@/stores';

type TextVariant = 'heading' | 'title' | 'subTitle' | 'description' | 'default';

const textBasicStyle: TextStyle = {
  textAlign: 'left',
  textAlignVertical: 'center',
  letterSpacing: letterSpacings.base,
};

export interface TextProps extends RNTextProps {
  children: string | ReactNode;
  variant?: TextVariant;
  fontSize?: number | string;
  fontWeight?: number | string;
  lineHeight?: number;
  textAlign?: string;
  color?: ColorValue;
  style?: StyleProp<TextStyle>;
}

const Text = ({
  children,
  variant = 'default',
  textAlign,
  fontSize,
  fontWeight,
  lineHeight,
  color,
  style,
  ...props
}: TextProps) => {
  const {theme} = useThemeStore();

  const {fonts, text} = theme;

  const textVariant = useMemo(
    () =>
      StyleSheet.create({
        heading: {
          fontFamily: fonts?.primary?.bold || fonts.default.bold,
          color: text.default,
          fontWeight: fontWeights.bold,
          fontSize: fontSizes.lg,
        },

        title: {
          fontFamily: fonts?.secondary?.medium || fonts.default.medium,
          fontSize: fontSizes.md,
          color: text.default,
          fontWeight: fontWeights.bold,
        },

        subTitle: {
          fontFamily: fonts?.tertiary?.regular || fonts.default.regular,
          fontSize: fontSizes.sm,
          fontWeight: fontWeights.bold,
          color: text.default,
        },

        description: {
          fontFamily: fonts?.primary?.regular,
          color: text.primary,
          fontSize: fontSizes.tiny,
        },

        default: {
          fontFamily: fonts?.primary?.regular,
          fontSize: fontSizes.xs,
          color: text.quaternary,
        },
      }),
    [fonts, text],
  );

  return (
    <RNText
      style={[
        textBasicStyle,
        textVariant[variant],
        {
          ...(color && {color}),
          ...(fontSize && {fontSize}),
          ...(fontWeight && {fontWeight}),
          ...(lineHeight && {lineHeight}),
          ...(textAlign && {textAlign}),
        } as TextStyle,
        style,
      ]}
      {...props}>
      {children}
    </RNText>
  );
};

export default memo(Text);
