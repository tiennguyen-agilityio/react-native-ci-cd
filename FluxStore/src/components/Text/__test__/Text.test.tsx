import {render, screen} from '@testing-library/react-native';
import Text from '..';
import {colors, fontSizes, fontWeights, lineHeights} from '@/themes';
import {useThemeStore} from '@/stores';

jest.mock('@/stores/theme.ts', () => ({
  useThemeStore: jest.fn(() => ({
    theme: {
      fonts: {
        primary: {
          bold: 'Primary-Bold',
          regular: 'Primary-Regular',
        },
        secondary: {medium: 'Secondary-Medium'},
        tertiary: {regular: 'Tertiary-Regular'},
        default: {
          bold: 'Default-Bold',
          medium: 'Default-Medium',
          regular: 'Default-Regular',
        },
      },
      text: {
        default: '#000000',
        primary: '#111111',
        quaternary: '#222222',
      },
    },
  })),
}));

describe('Text', () => {
  const props = {
    children: 'Lorem ipsum dolor sit amet',
  };
  it('renders default', () => {
    const {toJSON} = render(<Text {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with variant heading', () => {
    const {toJSON} = render(<Text {...props} variant="heading" />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with variant title', () => {
    const {toJSON} = render(<Text {...props} variant="title" />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with variant subTitle', () => {
    const {toJSON} = render(<Text {...props} variant="subTitle" />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with variant description', () => {
    const {toJSON} = render(<Text {...props} variant="description" />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with custom color, fontsize', () => {
    const {toJSON} = render(
      <Text
        {...props}
        color={colors.green[500]}
        fontSize={fontSizes['3xl']}
        fontWeight={fontWeights.medium}
        lineHeight={lineHeights.lg}
        textAlign="left"
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('falls back to default font if fonts.primary.bold is missing', () => {
    (useThemeStore as unknown as jest.Mock).mockReturnValueOnce({
      theme: {
        fonts: {
          default: {bold: 'Default-Bold'},
        },
        text: {
          default: '#000000',
        },
      },
    });

    render(<Text variant="heading">Fallback Font</Text>);
    const element = screen.getByText('Fallback Font');

    const fontFamilyStyle = (element.props.style || []).find((s: any) => s && s.fontFamily);

    expect(fontFamilyStyle).toBeDefined();
  });
});
