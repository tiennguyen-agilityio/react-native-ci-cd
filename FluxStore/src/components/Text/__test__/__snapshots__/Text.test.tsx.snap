// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders default 1`] = `
<Text
  style={
    [
      {
        "letterSpacing": 0,
        "textAlign": "left",
        "textAlignVertical": "center",
      },
      {
        "color": "#222222",
        "fontFamily": "Primary-Regular",
        "fontSize": 14,
      },
      {},
      undefined,
    ]
  }
>
  Lorem ipsum dolor sit amet
</Text>
`;

exports[`Text renders with custom color, fontsize 1`] = `
<Text
  style={
    [
      {
        "letterSpacing": 0,
        "textAlign": "left",
        "textAlignVertical": "center",
      },
      {
        "color": "#222222",
        "fontFamily": "Primary-Regular",
        "fontSize": 14,
      },
      {
        "color": "#5ece7b",
        "fontSize": 40,
        "fontWeight": 500,
        "lineHeight": 34,
        "textAlign": "left",
      },
      undefined,
    ]
  }
>
  Lorem ipsum dolor sit amet
</Text>
`;

exports[`Text renders with variant description 1`] = `
<Text
  style={
    [
      {
        "letterSpacing": 0,
        "textAlign": "left",
        "textAlignVertical": "center",
      },
      {
        "color": "#111111",
        "fontFamily": "Primary-Regular",
        "fontSize": 12,
      },
      {},
      undefined,
    ]
  }
>
  Lorem ipsum dolor sit amet
</Text>
`;

exports[`Text renders with variant heading 1`] = `
<Text
  style={
    [
      {
        "letterSpacing": 0,
        "textAlign": "left",
        "textAlignVertical": "center",
      },
      {
        "color": "#000000",
        "fontFamily": "Primary-Bold",
        "fontSize": 24,
        "fontWeight": 700,
      },
      {},
      undefined,
    ]
  }
>
  Lorem ipsum dolor sit amet
</Text>
`;

exports[`Text renders with variant subTitle 1`] = `
<Text
  style={
    [
      {
        "letterSpacing": 0,
        "textAlign": "left",
        "textAlignVertical": "center",
      },
      {
        "color": "#000000",
        "fontFamily": "Tertiary-Regular",
        "fontSize": 16,
        "fontWeight": 700,
      },
      {},
      undefined,
    ]
  }
>
  Lorem ipsum dolor sit amet
</Text>
`;

exports[`Text renders with variant title 1`] = `
<Text
  style={
    [
      {
        "letterSpacing": 0,
        "textAlign": "left",
        "textAlignVertical": "center",
      },
      {
        "color": "#000000",
        "fontFamily": "Secondary-Medium",
        "fontSize": 20,
        "fontWeight": 700,
      },
      {},
      undefined,
    ]
  }
>
  Lorem ipsum dolor sit amet
</Text>
`;
